File: 203. Remove Linked List Elements/203.RemoveLinkedListElements.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */

/**
 * Recursively remove all elements from the linked list that have the value val.
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var removeElements = function (head, val) {
  if (!head) return null; // Base case: if the list is empty, return null

  // Recurse down the list
  head.next = removeElements(head.next, val);

  // Decide whether to exclude the current head
  return head.val === val ? head.next : head;
};

File: 203. Remove Linked List Elements/203.RemoveLinkedListElements.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

public class Solution {
    /**
     * Recursively remove all nodes with value val from the linked list.
     * @param head Current head of the linked list
     * @param val Value to be removed
     * @return Updated head of the linked list
     */
    public ListNode removeElements(ListNode head, int val) {
        if (head == null) return null; // Base case: end of list

        // Recursively process the rest of the list
        head.next = removeElements(head.next, val);

        // Return next node if current node needs to be removed
        return head.val == val ? head.next : head;
    }
}